airflow-core/src/airflow/models/deadline.py

```diff
@@ -90,3 +93,77 @@ def _determine_resource() -> tuple[str, str]:
     def add_deadline(cls, deadline: Deadline, session: Session = NEW_SESSION):
         """Add the provided deadline to the table."""
         session.add(deadline)
+
+
+class DeadlineReference(Enum):
+    """
+    Store the calculation methods for the various Deadline Alert triggers.
+
+    TODO:  PLEASE NOTE This class is a placeholder and will be expanded in the next PR.
+
+    ------
+    Usage:
+    ------
+
+    Example use when defining a deadline in a DAG:
+
+    DAG(
+        dag_id='dag_with_deadline',
+        deadline=DeadlineAlert(
+            reference=DeadlineReference.DAGRUN_LOGICAL_DATE,
+            interval=timedelta(hours=1),
+            callback=hello_callback,
+        )
+    )
+
+    To parse the deadline reference later we will use something like:
+
+    dag.deadline.reference.evaluate_with(dag_id=dag.dag_id)
+    """
+
+    DAGRUN_LOGICAL_DATE = "dagrun_logical_date"
+
+
+class DeadlineAlert(LoggingMixin):
+    """Store Deadline values needed to calculate the need-by timestamp and the callback information."""
+
+    def __init__(
+        self,
+        reference: DeadlineReference,
+        interval: timedelta,
+        callback: Callable | str,
+        callback_kwargs: dict | None = None,
+    ):
+        super().__init__()
+        self.reference = reference
+        self.interval = interval
+        self.callback_kwargs = callback_kwargs
+        self.callback = self.get_callback_path(callback)
+
+    @staticmethod
+    def get_callback_path(_callback: str | Callable) -> str:
+        if callable(_callback):
+            # Get the reference path to the callable in the form `airflow.models.deadline.get_from_db`
+            return f"{_callback.__module__}.{_callback.__qualname__}"
+
+        # Check if the dotpath can resolve to a callable; store it or raise a ValueError
+        try:
+            _callback_module, _callback_name = _callback.rsplit(".", 1)
+            getattr(sys.modules[_callback_module], _callback_name)
+            return _callback
+        except (KeyError, AttributeError):
+            # KeyError if the path is not valid
+            # AttributeError if the provided value can't be rsplit
+            raise ValueError("callback is not a path to a callable")
```

Comment:
nit:
include the original error stack trace?
```suggestion
        except (KeyError, AttributeError) as e:
            # KeyError if the path is not valid
            # AttributeError if the provided value can't be rsplit
            raise ValueError(f"callback {callback} is not a path to a callable: {e}")
```

Reply:
Yup, good call.   FWIW, we can't use f-strings in error messages because Reasons, but I'll add the context as you suggested.

Reply:
I lied, it's log messages we can't use f-strings.   I was distracted, sorry about that!

Reply:
Going to resolve this one and continue the discussion in the comment below.